File: 05_fp_challenges/05_valid_parentheses/Solution05.hs
-- Valid Parentheses  (LeetCode 20)
-- =================================
-- Given a string of just ()[]{}, return True iff every bracket is closed by the
-- matching type in the correct order.
--   isValid "()[]{}" == True
--   isValid "(]" == False
--   isValid "([)]" == False
--   isValid "{[]}" == True

module Main where

isValid :: String -> Bool
isValid s = ans && length tmp_stack == 0
  where
    (ans, tmp_stack) =
      foldl
        ( \(bool, stack) x ->
            if bool == False
              then (False, stack)
              else case x of
                '(' -> (bool, x : stack)
                ')' -> if head stack == '(' then (bool, tail stack) else (False, stack)
                '{' -> (bool, x : stack)
                '}' -> if head stack == '{' then (bool, tail stack) else (False, stack)
                '[' -> (bool, x : stack)
                ']' -> if head stack == '[' then (bool, tail stack) else (False, stack)
                _ -> (False, stack)
        )
        (True, [])
        s

-- ---- tests ----
main :: IO ()
main = do
  check (isValid "()") True
  check (isValid "()[]{}") True
  check (isValid "(]") False
  check (isValid "([)]") False
  check (isValid "{[]}") True
  check (isValid "(") False
  check (isValid "") True
  putStrLn "all passed"
  where
    check got want
      | got == want = putStrLn ("  PASS  " ++ show got)
      | otherwise = error ("  FAIL  got " ++ show got ++ " want " ++ show want)

File: 05_fp_challenges/05_valid_parentheses/Solution05.java
// Valid Parentheses  (LeetCode 20)
// =================================
// Given a string of just ()[]{}, return true iff every bracket is closed by the
// matching type in the correct order.
//   isValid("()[]{}") == true
//   isValid("([)]") == false
//   isValid("{[]}") == true

import java.util.ArrayDeque;
import java.util.Deque;

public class Solution05 {

    static boolean isValid(String s) {
        throw new UnsupportedOperationException("TODO");
    }

    // ---- tests ----
    public static void main(String[] args) {
        check(isValid("()"),     true);
        check(isValid("()[]{}"), true);
        check(isValid("(]"),     false);
        check(isValid("([)]"),   false);
        check(isValid("{[]}"),   true);
        check(isValid("("),      false);
        check(isValid(""),       true);
        System.out.println("all passed");
    }

    static void check(boolean got, boolean want) {
        if (got != want) throw new AssertionError("FAIL got " + got + " want " + want);
        System.out.println("  PASS  " + got);
    }
}
